### PR TITLE
Add gui capabilities

### PR DESCRIPTION
**Pull request by ostenvighx**
Adds a gui capability and interface, IGuiProvider and CapabilityGuiProvider.

This also adds base implementations for TileEntities (GuiProviderTile), Itemstacks (GuiProviderItem), and Entities (GuiProviderEntity) as well as a blank implementation to extend on.

The objective here is to push the gui handling code closer to the actual objects they go to. You can get an instance of a gui straight from the block it belongs to, or right from an entity. In addition, this enables implementations to add additional data to the gui opening packet by overriding IGuiProvider's serialize and deserialize methods.

### CODE REVIEW

File: src/main/java/net/minecraftforge/fml/common/network/internal/FMLMessage.java [hunk 1]
```diff
@@ -66,9 +66,7 @@ void toBytes(ByteBuf buf)
         int windowId;
         String modId;
         int modGuiId;
-        int x;
-        int y;
-        int z;
+        int x, y, z;
```
Comment (williewillus) [inline src/main/java/net/minecraftforge/fml/common/network/internal/FMLMessage.java]: this change is completely unnecessary

File: src/main/java/net/minecraftforge/gui/ForgeGuiHandler.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import net.minecraftforge.fml.relauncher.Side;
+import net.minecraftforge.gui.capability.IGuiProvider;
+
+public class ForgeGuiHandler
+{
+    public static void openGui(EntityPlayer player, World worldObj, IGuiProvider provider)
+    {
+        if(provider == null) return;
+        Object owner = provider.getOwner(player, worldObj);
+        if(FMLCommonHandler.instance().getEffectiveSide() == Side.CLIENT)
```
Comment (williewillus) [inline src/main/java/net/minecraftforge/gui/ForgeGuiHandler.java]: if you're going to use getEffectiveSide then might as well check world.isRemote. Same semantics, but second is more reliable

Comment (ostenvighx) [inline src/main/java/net/minecraftforge/gui/ForgeGuiHandler.java]: Used because original network handler uses getEffectiveSide.

File: src/main/java/net/minecraftforge/fml/common/network/internal/FMLMessage.java (hunk 1 above)
Comment (ostenvighx) [inline src/main/java/net/minecraftforge/fml/common/network/internal/FMLMessage.java]: Makes the code more readable and reduces line count slightly?

Comment (mezz) [inline src/main/java/net/minecraftforge/fml/common/network/internal/FMLMessage.java]: The fewer changes you make, the fewer things people have to complain about.

**Comment by mezz** [comment]
This looks very promising but it has formatting and other issues, I'm working with ostenvighx on IRC to help fix it up.

File: src/main/java/net/minecraftforge/gui/ForgeGuiHandler.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+        Object owner = provider.getOwner(player, worldObj);
+        if (worldObj.isRemote)
+        {
+            Object gui = provider.getClientGuiElement(player, worldObj, owner);
+            FMLCommonHandler.instance().showGuiScreen(gui);
+            return;
+        }
+
+        if (player instanceof FakePlayer || player instanceof EntityPlayerSP) return;
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/gui/ForgeGuiHandler.java]: `EntityPlayerSP` does not exist on a server.

File: src/main/java/net/minecraftforge/gui/OpenGuiMessage.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import net.minecraft.client.multiplayer.WorldClient;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.util.IThreadListener;
+import net.minecraftforge.fml.client.FMLClientHandler;
+import net.minecraftforge.fml.common.FMLCommonHandler;
+import net.minecraftforge.fml.common.network.NetworkRegistry;
+import net.minecraftforge.fml.common.network.simpleimpl.IMessage;
+import net.minecraftforge.gui.capability.IGuiProvider;
+
+public class OpenGuiMessage implements IMessage
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/gui/OpenGuiMessage.java]: FMLs channel is not using the `IMessage` system.

File: src/main/java/net/minecraftforge/gui/capability/IGuiProvider.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+    /**
+     * Gets the gui provider object from the world, player, etc.
+     * Used in the get*Element methods to pass a direct instance to them.
+     *
+     * @param player The player instance.
+     * @param world The world.
+     * @return
+     */
+    Object getOwner(EntityPlayer player, World world);
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/gui/capability/IGuiProvider.java]: The statefulness introduced here (`deserialize` and `getOwner` must be called after each other!) is not very nice imho.

File: src/main/java/net/minecraftforge/gui/capability/impl/GuiProviderItem.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.EnumHand;
+import net.minecraft.world.World;
+import net.minecraftforge.fml.common.network.ByteBufUtils;
+import net.minecraftforge.gui.capability.IGuiProvider;
+
+public abstract class GuiProviderItem implements IGuiProvider
+{
+    private ItemStack stack;
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/gui/capability/impl/GuiProviderItem.java]: Why is this here? It's never used.

File: src/main/java/net/minecraftforge/gui/capability/impl/GuiProviderTile.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+import net.minecraftforge.gui.capability.IGuiProvider;
+
+public abstract class GuiProviderTile implements IGuiProvider
+{
+    private BlockPos position;
+    private EnumFacing side;
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/gui/capability/impl/GuiProviderTile.java]: Same as above, never used.

File: patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch [hunk 8]
```diff
@@ -443,6 +443,8 @@
 +
 +
 +    /* ======================================== FORGE START =====================================*/
++    public void openGui(net.minecraftforge.gui.capability.IGuiProvider provider) { net.minecraftforge.gui.ForgeGuiHandler.openGui(this, field_70170_p, provider); }
```
Comment (mezz) [inline patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch]: You can expand this to multiple lines so it's clear, it's in the Forge section. As long as you're not importing anything it'll be fine.

File: src/main/java/net/minecraftforge/gui/capability/IGuiProvider.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    /**
+     * Gets a server-side container to handle slots and interaction on the server.
+     *
+     * @param player The player viewing the gui.
+     * @param world The world the gui is being accessed on.
+     * @param owner The owner object. Usually a tile, item, or entity.
+     * @return A Container the server uses for processing.
+     */
+    @Nullable
+    Object getServerGuiElement(EntityPlayer player, World world, @Nullable Object owner);
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/gui/capability/IGuiProvider.java]: Why does this not return `Container`?

File: src/main/java/net/minecraftforge/gui/capability/CapabilityGuiProvider.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+import net.minecraftforge.common.capabilities.CapabilityManager;
+import net.minecraftforge.gui.capability.impl.GuiProviderBlank;
+
+import java.util.concurrent.Callable;
+
+public class CapabilityGuiProvider
+{
+
+    @CapabilityInject(IGuiProvider.class)
+    public static Capability<IGuiProvider> GUI_PROVIDER_CAPABILITY;
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/gui/capability/CapabilityGuiProvider.java]: `final`

File: src/main/java/net/minecraftforge/gui/OpenGuiMessage.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+                        Handler.this.process(msg);
+                    }
+                });
+            }
+        }
+
+        private void process(OpenGuiMessage msg)
+        {
+            EntityPlayer player = FMLClientHandler.instance().getClientPlayerEntity();
+            WorldClient world = FMLClientHandler.instance().getWorldClient();
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/gui/OpenGuiMessage.java]: The server does not know `WorldClient`.

File: src/main/java/net/minecraftforge/gui/OpenGuiMessage.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+            Object owner = msg.provider.getOwner(player, world);
+
+            if(owner == null) return;
+            Object gui = msg.provider.getClientGuiElement(player, world, owner);
+            FMLClientHandler.instance().showGuiScreen(gui);
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/gui/OpenGuiMessage.java]: The server does not know `FMLClientHandler`. You are meant to use `FMLCommonHandler`, which delegates to client or serer.

File: src/main/java/net/minecraftforge/gui/ForgeGuiHandler.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        EmbeddedChannel embeddedChannel = FMLNetworkHandler.getServerChannel();
+        embeddedChannel.attr(FMLOutboundHandler.FML_MESSAGETARGET).set(FMLOutboundHandler.OutboundTarget.PLAYER);
+        embeddedChannel.attr(FMLOutboundHandler.FML_MESSAGETARGETARGS).set(playerMP);
+        embeddedChannel.writeOutbound(openGui);
+
+        playerMP.openContainer = container;
+        playerMP.openContainer.windowId = windowId;
+        playerMP.openContainer.addListener(playerMP);
+
+        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.PlayerContainerEvent.Open(player, player.openContainer));
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/gui/ForgeGuiHandler.java]: Why? This is not a patch to vanilla.

**Comment by Actuarius** [comment]
@tterrag1098 added labels [Feature]

File: src/main/java/net/minecraftforge/gui/ForgeGuiHandler.java (hunk 13 above)
Comment (ostenvighx) [inline src/main/java/net/minecraftforge/gui/ForgeGuiHandler.java]: See the original code in FML's network handler. Duplicated from there.

File: src/main/java/net/minecraftforge/gui/OpenGuiMessage.java (hunk 12 above)
Comment (ostenvighx) [inline src/main/java/net/minecraftforge/gui/OpenGuiMessage.java]: This is only run on the client, and the FML networking gui code references FMLClientHandler in their handling method too. Copied from there.

File: src/main/java/net/minecraftforge/fml/common/network/IGuiHandler.java [hunk 14]
```diff
@@ -22,6 +22,7 @@
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.world.World;
 
+@Deprecated // Use capability version instead
```
Comment (mezz) [inline src/main/java/net/minecraftforge/fml/common/network/IGuiHandler.java]: Use a  javadoc here, with a link to the new thing.
```java
/**
 * @deprecated, use {@link net.minecraftforge.gui.capability.IGuiProvider} instead
 */
@Deprecated
```

File: src/main/java/net/minecraftforge/gui/ForgeGuiHandler.java (hunk 3 above)
Comment (mezz) [inline src/main/java/net/minecraftforge/gui/ForgeGuiHandler.java]: I think you should rearrange this method from here down to avoid early returns. 
Then you can have
```java
if (player instanceof EntityPlayerMP && !(player instanceof FakePlayer)) {
   ...
}
```

File: src/main/java/net/minecraftforge/gui/capability/CapabilityGuiProvider.java [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+
+import java.util.concurrent.Callable;
+
+public class CapabilityGuiProvider
+{
+
+    @CapabilityInject(IGuiProvider.class)
+    public final static Capability<IGuiProvider> GUI_PROVIDER_CAPABILITY = null;
+
+    public static void register(){
```
Comment (mezz) [inline src/main/java/net/minecraftforge/gui/capability/CapabilityGuiProvider.java]: brace should be on new line

File: src/main/java/net/minecraftforge/gui/capability/IGuiProvider.java [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+     * @return
+     */
+    T getOwner(EntityPlayer player, World world);
+
+    /**
+     * Gets a client-side gui element (may be null) to display to a player.
+     *
+     * @param player The player viewing the gui.
+     * @param world The world the gui is being displayed in.
+     * @param owner The owner object of the gui. Usually a tile, item, or entity.
```
Comment (mezz) [inline src/main/java/net/minecraftforge/gui/capability/IGuiProvider.java]: Please explain the `@Nullable` by describing the conditions where owner can be null

File: src/main/java/net/minecraftforge/gui/capability/IGuiProvider.java [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+     */
+    @Nullable
+    Object getClientGuiElement(EntityPlayer player, World world, @Nullable T owner);
+
+    /**
+     * Gets a server-side container to handle slots and interaction on the server.
+     *
+     * @param player The player viewing the gui.
+     * @param world The world the gui is being accessed on.
+     * @param owner The owner object. Usually a tile, item, or entity.
```
Comment (mezz) [inline src/main/java/net/minecraftforge/gui/capability/IGuiProvider.java]: Please explain the `@Nullable` by describing the conditions where owner can be null

File: src/main/java/net/minecraftforge/gui/capability/IGuiProvider.java [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * Gets the gui provider object from the world, player, etc.
+     * Used in the get*Element methods to pass a direct instance to them.
+     *
+     * @param player The player instance.
+     * @param world The world.
+     * @return
+     */
+    T getOwner(EntityPlayer player, World world);
```
Comment (mezz) [inline src/main/java/net/minecraftforge/gui/capability/IGuiProvider.java]: Is this return value also `@Nullable`?

File: src/main/java/net/minecraftforge/gui/capability/impl/GuiProviderItem.java [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        this.stack = stack;
+        this.hand = hand;
+    }
+
+    @Override
+    public void deserialize(ByteBuf buffer)
+    {
+        stack = ByteBufUtils.readItemStack(buffer);
+        hand = buffer.readInt() == 0 ? EnumHand.MAIN_HAND : EnumHand.OFF_HAND;
```
Comment (mezz) [inline src/main/java/net/minecraftforge/gui/capability/impl/GuiProviderItem.java]: You should encode this as a boolean instead of an int.

File: src/main/java/net/minecraftforge/gui/capability/impl/GuiProviderItem.java [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    @Override
+    public void serialize(ByteBuf buffer)
+    {
+        ByteBufUtils.writeItemStack(buffer, stack);
+        buffer.writeInt(hand == EnumHand.MAIN_HAND ? 0 : 1);
```
Comment (mezz) [inline src/main/java/net/minecraftforge/gui/capability/impl/GuiProviderItem.java]: You should encode this as a boolean instead of an int.

File: src/main/java/net/minecraftforge/gui/capability/impl/GuiProviderBlank.java [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+    @Override
+    public void deserialize(ByteBuf buffer) { }
+
+    @Override
+    public void serialize(ByteBuf buffer) { }
+
+    @Override
+    public Object getOwner(EntityPlayer player, World world)
+    {
+        return null;
```
Comment (mezz) [inline src/main/java/net/minecraftforge/gui/capability/impl/GuiProviderBlank.java]: Maybe you should make this abstract instead of returning null, since the method is not marked `@Nullable`?

File: src/test/java/net/minecraftforge/test/TestGuiCapability.java [hunk 22]
```diff
@@ -0,0 +1,308 @@
...
+            setUnlocalizedName("gui.item");
+            setCreativeTab(CreativeTabs.MISC);
+            setMaxStackSize(1);
+            GameRegistry.register(this);
+        }
+
+        @Override
+        public ActionResult<ItemStack> onItemRightClick(ItemStack itemStackIn, World worldIn, EntityPlayer playerIn, EnumHand hand)
+        {
+            if (itemStackIn.hasCapability(CapabilityGuiProvider.GUI_PROVIDER_CAPABILITY, EnumFacing.UP))
```
Comment (mezz) [inline src/test/java/net/minecraftforge/test/TestGuiCapability.java]: I still think using `UP` or the hand for the `EnumFacing` is fishy here. Nobody else is going to know how to open your gui. I think `null` should be used, unless you know which side of the block is being clicked.

**Comment by ostenvighx** [comment]
The test mod only shows an example. It could be null where it says EnumFacing.UP, it really doesn't matter.

File: patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch [hunk 23]
```diff
@@ -443,6 +443,11 @@
 +
 +
 +    /* ======================================== FORGE START =====================================*/
```
Comment (loordgek) [inline patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch]: do we still need the other opengui ?
https://github.com/ostenvighx/MinecraftForge/blob/126bffadc41ba5e1564f772eb2a215523756b229/patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch#L428-L442

Comment (ostenvighx) [inline patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch]: Yes, we keep the other version for backwards compatibility. I should mark it deprecated, though.

Comment (raoulvdberge) [inline patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch]: No, you shouldn't. Implementing hard interfaces on tiles instead of using tile capabilities isn't deprecated either.

**Comment by LexManos** [comment]
What does this provide that the old system does not? Why would we need this? 
Also from a quick glance this seems fundamentally broken as there is no way to get the correct provider on the client side. So it won't work for anything except single player.

**Comment by ostenvighx** [comment]
> What does this provide, and why would we need this?

This provides a way for gui providers to pass additional data along in the open gui packet if it's needed, as well as providing a means of getting another mod's gui easily. (EnderIO remote blocks, for example, could take advantage of this)

> Doesn't cleanup the code

Okay, I might give you this one. It's a bit more boilerplate code, but it removes the need to register a gui handler explicitly, removes the giant switch statement and the extra data (you only pass what you need in the serialization methods), and it also allows people to attach guis to things. The test mod opens a gui when you hit a pig.

File: src/main/java/net/minecraftforge/gui/capability/IGuiProvider.java [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+     */
+    IGuiProvider deserialize(ByteBuf buffer);
+
+    /**
+     * Serializes data on the server to send along with the open gui packet.
+     * Use this to add any information you need to open the gui properly on the client.
+     *
+     * @param buffer
+     */
+    void serialize(ByteBuf buffer, Object... extras);
```
Comment (mezz) [inline src/main/java/net/minecraftforge/gui/capability/IGuiProvider.java]: Needs javadoc to explain `extras`

File: src/main/java/net/minecraftforge/gui/capability/impl/GuiProviderBase.java [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+        this.owner = null;
+    }
+
+    public GuiProviderBase(T in)
+    {
+        this.owner = in;
+    }
+
+    @Override
+    public IGuiProvider deserialize(ByteBuf buffer) { return null; }
```
Comment (mezz) [inline src/main/java/net/minecraftforge/gui/capability/impl/GuiProviderBase.java]: should it return `this`?

**Comment by LexManos** [comment]
>This provides a way for gui providers to pass additional data along in the open gui packet if it's needed, 

This is the only legitimate thing your PR has going for it. And this could EASILY be added onto the system we already have in place.

>as well as providing a means of getting another mod's gui easily. (EnderIO remote blocks, for example, could take advantage of this)

Why would you use this.. It also doesn't deal with the range restrictions on most guis. Nore doesnt to really provide a generic way to get the UI.

>It's a bit more boilerplate code, 

Quite a bit
>but it removes the need to register a gui handler explicitly,

No it doesn't it just moves it to the instances instead of the central place, so in fact it means you have to register MORE 
>removes the giant switch statement

Thats just your implementation you could do it in a million different ways, however a switch or lookup table is honestly the best way.
>and the extra data (you only pass what you need in the serialization methods),

What extra data?
>and it also allows people to attach guis to things. The test mod opens a gui when you hit a pig.

You can already do this in the existing system. This really doesn't allow for attaching a GUI to anything. It just changes where you implement the 'open' method and makes it severly more complicated to open the GUI.

So far, I have not seen anything from this PR that is worth while. The only legit thing you've brought up is a possible need for 'extra data' in the open packet. If you could justify that part we may be able to make something to suit your needs.

**Comment by blay09** [comment]
> Why would you use this.. It also doesn't deal with the range restrictions on most guis. Nore doesnt to really provide a generic way to get the UI.

I have a shared GUI implemented by my mod that has a [go-back](http://media-elerium.cursecdn.com/attachments/64/438/filterpreview.png) button in the right corner to go back to the original GUI, which is provided by the mod implementing my API. Right now those mods need to [register a callback function](https://github.com/blay09/RefinedRelocation2/blob/1.10/src/api/java/net/blay09/mods/refinedrelocation/api/container/ITileGuiHandler.java) for that, so having a standardized way of getting a tile's GUI would have made that a lot easier.

**Comment by ostenvighx** [comment]
Closing this and re-opening after a discussion that capabilities for this just won't work, because internal handling won't be supported (extra data being passed kinda kills that)
